x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java

```diff
@@ -465,19 +464,22 @@ public void datafeedTimingStats(List<String> jobIds, Consumer<Map<String, Datafe
                         String jobId = jobIds.get(i);
                         MultiSearchResponse.Item itemResponse = msearchResponse.getResponses()[i];
                         if (itemResponse.isFailure()) {
-                            errorHandler.accept(itemResponse.getFailure());
+                            listener.onFailure(itemResponse.getFailure());
+                            return;
                         } else {
                             SearchResponse searchResponse = itemResponse.getResponse();
                             ShardSearchFailure[] shardFailures = searchResponse.getShardFailures();
                             int unavailableShards = searchResponse.getTotalShards() - searchResponse.getSuccessfulShards();
                             if (shardFailures != null && shardFailures.length > 0) {
                                 LOGGER.error("[{}] Search request returned shard failures: {}", jobId, Arrays.toString(shardFailures));
-                                errorHandler.accept(
+                                listener.onFailure(
                                     new ElasticsearchException(ExceptionsHelper.shardFailuresToErrorMsg(jobId, shardFailures)));
+                                return;
```

Comment:
Same as the my later comment. As we return now, we can simplify by not using `else`.